Windows support fixes

wesm opened the pull request:
First attempt at the suggested fixes

cc @willyd @jakirkham @maxris 

Closes #4

willyd committed:
Moved to version 2.2.0 and added windows support

willyd committed:
Fixed lint issue

willyd committed:
Disabled build for python 3.6 and fixed tests on windows

willyd committed:
Rerendered with conda-smithy 2.0.1

willyd committed:
Skip unsupported vc version
Renamed build folder to avoid conflict with the bazel BUILD file

wesm committed:
Patch CMakeLists.txt to install libraries in bindir

wesm committed:
Remove existing post-install patch

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

wesm committed:
Check in bld.bat changes

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,38 @@ package:
+  features:
+    - vc9   # [win and py27]
+    - vc10  # [win and py34]
+    - vc14  # [win and py35]
```
Should add a feature for Windows Python 3.6 for good measure (even though we are skipping it for now).

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,38 @@ package:
+  requires:
+    - python 2.7.*  # [win and py27]
+    - python 3.4.*  # [win and py34]
+    - python 3.5.*  # [win and py35]
```
Should add Python 3.6 for good measure even though we are skipping it for now.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -7,22 +7,38 @@ package:
     - exit $(test -f ${PREFIX}/lib/libgflags_nothreads.dylib)  # [osx]
     - exit $(test -f ${PREFIX}/lib/libgflags.so)               # [linux]
     - exit $(test -f ${PREFIX}/lib/libgflags_nothreads.so)     # [linux]
+    - if not exist %PREFIX%\\Library\\bin\\gflags.dll exit 1   # [win]
```
Should we be testing for a `.lib` file as well?

willyd replied:
Makes sense. We should test for it.

jakirkham commented on recipe/windows_cmake_install_bindir.patch:
```diff
@@ -0,0 +1,13 @@
+ if (OS_WINDOWS)
+   set (RUNTIME_INSTALL_DIR Bin)
+-  set (LIBRARY_INSTALL_DIR Lib)
++  set (LIBRARY_INSTALL_DIR Bin)
```
Not totally sure that we need this patch as noted in this [comment]( https://github.com/conda-forge/gflags-feedstock/pull/4#issuecomment-292349781 ).

willyd replied:
I confirm that with conda 4.2.13 only bin is added to the path.

jakirkham replied:
@msarahan, could you please take a look at this patch? I'd appreciate your feedback on it. Particularly whether this is necessary on Windows or not.

jakirkham replied:
Maybe we need to set the `ARCHIVE_INSTALL_DIR` to `Lib`.

willyd replied:
Actually we need to change the `install` command since it installs everything in `LIBRARY_INSTALL_DIR`. 

I think this should do it:
```
install(TARGETS <library_name>
        RUNTIME DESTINATION ${RUNTIME_INSTALL_DIR}
        LIBRARY DESTINATION ${LIBRARY_INSTALL_DIR}
        ARCHIVE DESTINATION ${LIBRARY_INSTALL_DIR})
```

see https://github.com/gflags/gflags/blob/v2.2.0/CMakeLists.txt#L504

jakirkham replied:
Indeed. Could you please send an updated patch? I lack a Windows machine to test on. So this would be slow going for me.

willyd replied:
On it.

jakirkham replied:
Thanks.

willyd replied:
I just pushed a new commit with a new patch on my windows branch. It ended up in PR #4. Is there any way I can push to this one?

jakirkham commented on recipe/build.sh:
```diff
@@ -1,7 +1,7 @@
-mkdir build
-cd build
+mkdir build_release
+cd build_release
```
Is there any particular reason for renaming this?

willyd replied:
Yes. gflags includes a bazel BUILD file. Windows is case insensitive so the name conflicted.

jakirkham replied:
Right, makes sense. Mac has similar issues. So it is good to do the same thing here.

jakirkham reviewed (commented):
Thanks for cleaning this up a bit, @wesm. Looks nice. A few comments and questions. Some might be better addressed by @willyd directly.

jakirkham committed:
Merge remote-tracking branch 'conda-forge/master' into 'wesm/windows'

jakirkham committed:
MNT: Re-rendered with conda-smithy 2.2.2

jakirkham commented:
Made some changes to `master` to cleanup some things (this recipe is a bit old). Have resolved conflicts for this branch and pushed the changes to this PR.

jakirkham commented:
Seems VC 2008 is failing because of issue ( https://github.com/gflags/gflags/issues/197 ).

Edit: Would normally try adding `msinttypes` to `build` with a `py<35` selector, but it seems we did that already. 😕

willyd commented:
I think we need to add `INTTYPES_FORMAT=VC7` option and add the location of msintypes as include directories. I will give this a shot.

willyd commented:
Turns out conda adds the right include directories and that INTTYPES_FORMAT is not required. The package builds successfully for me on py 2.7. 

However, I had to switch to the Ninja generator (could have used nmake) since I don't have VS 2008 installed and only the VC C++ For Python 2.7.

@wesm the patch is in the right direction but it now installs both the .lib and .dll in `bin` whereas it should install the .lib in `lib`.

Output from my build
```
-- Installing: C:/conda-bld/gflags_1491585830643/_b_env/Library/Bin/gflags.lib
-- Installing: C:/conda-bld/gflags_1491585830643/_b_env/Library/Bin/gflags.dll
```

jakirkham committed:
Use NMake to build on Windows

Seem to be having some issues with the Visual Studio generators.
Particular when it comes to building with VS 2008. So this switches over
to NMake, which we have historically had more success with.

jakirkham commented:
Have pushed a change to this PR to switch to using the NMake generator on Windows. Based on builds from my fork, it seems this [works]( https://ci.appveyor.com/project/jakirkham/gflags-feedstock/build/1.0.1 ).

schuhschuh commented:
@willyd Would you mind sharing your findings of what is needed to resolve issue gflags/gflags#197 in the thread of the gflags project to help us fix it for everyone once and for all? I am not usually working on Windows, so it takes me longer to get around such issues. But it seems you guys had a look already.

> Turns out conda adds the right include directories and that INTTYPES_FORMAT is not required.

Which are these? What value does the CMake check pick up for `INTTYPES_FORMAT`? See [here](https://github.com/gflags/gflags/blob/4a694e87361d08eff5c4c9e9f551b1a0d41f7c40/CMakeLists.txt#L264).

willyd commented:
@schuhschuh From my CMakeCache.txt
```
//Format of integer types: "C99" (uint32_t), "BSD" (u_int32_t),
// "VC7" (__int32)
INTTYPES_FORMAT:STRING=VC7
```
Other than that we had to install [msinttypes](https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/msinttypes/msinttypes-r26.zip) for VS 2010 and lower.

schuhschuh commented:
Thanks, I take it the resolution is to have users of such old compilers install msinttypes then.

isuruf commented:
Or set this variable conditionally, https://github.com/gflags/gflags/blob/master/CMakeLists.txt#L227

schuhschuh commented:
Thanks, @isuruf. I've noted this was mentioned in a comment before. So I guess it would be good to add a configure time check for this header and always assume `INTTYPES_FORMAT=VC7` if not present.

jakirkham commented:
Yeah, one could use the [`CheckIncludeFile`]( https://cmake.org/cmake/help/v3.0/module/CheckIncludeFiles.html ) module to actually check for these headers.

schuhschuh commented:
Yes, that's been done for Unix systems already. I just thought it could be skipped for MSVC... anyway, already done: https://github.com/gflags/gflags/pull/224
